torch_em/transform/augmentation.py

```diff
@@ -90,32 +100,72 @@ def is_interpolatable(self, tensor):
-            augmentation.flags['interpolation'] = torch.tensor(resampler.value)
+            resampler = kornia.constants.Resample.get("BILINEAR" if interpolatable else "NEAREST")
+            augmentation.flags["interpolation"] = torch.tensor(resampler.value)
 
-        transformed = augmentation(tensor, params)
-        return transformed, augmentation._params
+    def _get_eye(self, tensor: torch.Tensor):
+        return kornia.eye_like(4 if self.is3D else 3, tensor)
 
-    def forward(self, *tensors):
+    def ensure_batch_tensor(self, tensor: torch.Tensor):
+        ensure_batch = (
+            kornia.augmentation.utils.helpers._transform_input3d
+            if self.is3D
+            else kornia.augmentation.utils.helpers._transform_input
+        )
+        return ensure_batch(ensure_tensor(tensor, self.dtype))
+
+    def forward(
+        self, *tensors, trans_matrices: Optional[Sequence[torch.Tensor]] = None
+    ) -> Union[List[torch.Tensor], Tuple[List[torch.Tensor], List[torch.Tensor]]]:
         interpolatable = [self.is_interpolatable(tensor) for tensor in tensors]
-        tensors = [ensure_tensor(tensor, self.dtype) for tensor in tensors]
-        for aug in self.augmentations:
+        tensors = [self.ensure_batch_tensor(tensor) for tensor in tensors]
+
+        if trans_matrices is None:
+            trans_matrices = [self._get_eye(t) for t in tensors]
+        else:
+            trans_matrices = list(trans_matrices)
 
-            t0, params = self.transform_tensor(aug, tensors[0], interpolatable[0])
-            transformed_tensors = [t0]
-            for tensor, interpolate in zip(tensors[1:], interpolatable[1:]):
-                tensor, _ = self.transform_tensor(aug, tensor, interpolate, params=params)
-                transformed_tensors.append(tensor)
+        assert len(trans_matrices) == len(tensors)
+        for ti, tensor in enumerate(tensors):
+            if trans_matrices[ti] is None:
+                trans_matrices[ti] = self._get_eye(tensor)
 
-            tensors = transformed_tensors
-        return tensors
+        transformed_tensors = []
+        all_params = [None] * len(self.augmentations)
+        for ti, (tensor, interpolate) in enumerate(zip(tensors, interpolatable)):
+            for ai, aug in enumerate(self.augmentations):
+                self._configure_augmentation(aug, interpolatable)
+                tensor, trans_matrices[ti] = aug((tensor, trans_matrices[ti]), all_params[ai])
+
+                if all_params[ai] is None:
+                    all_params[ai] = aug._params
+                else:
+                    assert all_params[ai] == aug._params
+
+            transformed_tensors.append(tensor)
+
+        return (transformed_tensors, trans_matrices) if self.return_transform else transformed_tensors
 
     def halo(self, shape):
         return self.halo
 
+    def apply_inverse(self, *tensors: torch.Tensor, forward_transforms: Sequence[torch.Tensor], padding_mode="border"):
```

Comment:
This only works for affine trafos / anything that can be expressed as affine trafo. But that's certainly not the case for all augmentations we have, e.g. elastic deformations.

Reply:
renamed it to `apply_inverse_affine` to make it explicit that it applies only to transformations that can be expressed as an affine transformation.